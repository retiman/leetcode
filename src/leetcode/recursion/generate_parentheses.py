# DIFFICULTY: MEDIUM
# ------------------
#
# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
#
# See https://leetcode.com/problems/generate-parentheses
class Solution:
    def generateParenthesis(self, n: int) -> list[int]:
        """
        SOLUTION
        --------

        A naive way to do this would be to do something like generate all the parentheses for n-1, then add more
        parentheses to each of the elements generated.  For example, by transforming each element of the n-1 solution to
        `${e}()`, `()${e}`, and `($e)`.

        This doesn't work for values like n=4, because you will miss solutions like `(())(())`.  Instead, we'll
        construct the string as we go along, keeping track of how many open and close parentheses we have used.  In
        total, the resulting strings will each have length n*2.

        This is because well formed strings have additional constraints; when generating a power set using the naive
        method, for example, you don't need to worry about balancing the elements within the sub set.  In contrast to
        generating a power set, you would have far fewer resulting elements.

        COMPLEXITY
        ----------

        Time complexity is uh... we'll just go with O(4^n) since we can generate all possible 2 * n length strings using
        characters '(' and ')'.  With 2 characters, it's 2^(2 * n) = 4^n.

        Space complexity is O(4^n).
        """
        if n == 0:
            return []

        result = []

        def generate(text, opens: int, closes: int):
            # The max length of the string is n * 2 because each open paren requires a close paren.  We can't get fewer
            # than n * 2 characters in a string either, because we have to generate all possible combinations.
            if len(text) == n * 2:
                result.append(text)
                return

            if opens < n:
                generate(text + "(", opens + 1, closes)

            if closes < opens:
                generate(text + ")", opens, closes + 1)

        generate("", 0, 0)
        return result
