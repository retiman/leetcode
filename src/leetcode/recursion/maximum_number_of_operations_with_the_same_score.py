# DIFFICULTY: MEDIUM
# ------------------
#
# Given an array of integers called nums, you can perform any of the following operation while nums contains at least
# 2 elements:
#
# - Choose the first two elements of nums and delete them.
# - Choose the last two elements of nums and delete them.
# - Choose the first and the last elements of nums and delete them.
#
# The score of the operation is the sum of the deleted elements.
#
# Your task is to find the maximum number of operations that can be performed, such that all operations have the same
# score.
#
# Return the maximum number of operations possible that satisfy the condition mentioned above.
#
# See https://leetcode.com/problems/maximum-number-of-operations-with-the-same-score-ii
class Solution:
    def maxOperations(self, nums: list[int]) -> int:
        """
        SOLUTION
        --------

        The range of possible scores is determined by applying the operation once.  Subsequent operations must produce
        the same score, so there are only three possible operations.  For each operation, we should apply them, then
        follow paths of operations that result in the same score until the array is exhausted.

        We COULD use the sliding window technique by applying each operation and then only following the operation paths
        that have the same score as the one we are considering.  However, this would only work if at each choice of
        three operations, the scores are unique.  If there are two or more choices with the same score, then we need to
        explore multiple operation paths at once.

        To explore multiple operation paths at once, we'll need to use a backtracking approach.

        COMPLEXITY
        ----------

        Time complexity is O(3^n) because we are exploring all possible operations.

        Space complexity is O(n) because we are storing the operations in stack frames.
        """
        if len(nums) < 2:
            return 0

        # These are the range of possible scores we can have as we truncate the array.
        scores: set[int] = set()
        scores.add(nums[0] + nums[1])
        scores.add(nums[0] + nums[-1])
        scores.add(nums[-1] + nums[-2])

        # There is an opportunity for optimization when using backtracking.  We can maintain a memoization table to keep
        # track of previously computed values.
        memo: dict[tuple[int, int, int], int] = {}

        # Adapt the sliding window approach so we can explore multiple paths at once by backtracking.
        def backtrack(score: int, left: int, right: int, paths: int) -> int:
            if left >= right:
                return paths

            # Check the memoization table to see if we've previously computed this value.
            if (score, left, right) in memo:
                return memo[(score, left, right)]

            # This is the local max count, after exploring all possible paths.
            local = paths

            # Check if the first 2 elements are equal to the score, and if the left and right indices are in range.
            #
            # To be in range, the array must at least look like this: [left, left + 1/right].
            if left + 1 <= right and nums[left] + nums[left + 1] == score:
                local = max(local, backtrack(score, left + 2, right, paths + 1))

            # Check if the last 2 elements are equal to the score, and if the left and right indices are in range.
            #
            # To be in range, the array must at least look like this: [left/right - 1, right]
            if left <= right - 1 and nums[right] + nums[right - 1] == score:
                local = max(local, backtrack(score, left, right - 2, paths + 1))

            # Check if the first and last elements are equal to the score, and if the left and right indices are in
            # range.
            #
            # To be in range, the array must at least look like this: [left, right]
            if left < right and nums[left] + nums[right] == score:
                local = max(local, backtrack(score, left + 1, right - 1, paths + 1))

            # Update the memoization table.
            memo[(score, left, right)] = local
            return local

        # For each score, check out how many times we can keep applying operations.
        xs = list(scores)
        maxes = [backtrack(x, 0, len(nums) - 1, 0) for x in xs]
        return max(maxes)
