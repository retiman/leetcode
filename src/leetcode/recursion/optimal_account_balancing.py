# DIFFICULTY: HARD
# ----------------
#
# You are given an array of transactions transactions where transactions[i] = [fromi, toi, amounti] indicates that the
# person with ID = fromi gave amounti $ to the person with ID = toi.
#
# Return the minimum number of transactions required to settle the debt.
#
# See https://leetcode.com/problems/optimal-account-balancing
import math


class Solution:
    def minTransfers(self, transactions: list[list[int]]) -> int:
        """
        SOLUTION
        --------

        We can model this problem as a graph where each person is a node and the amount they owe or are owed is the
        weight of the edge between them.  The goal is to find the minimum number of transactions to settle all debts.

        We can use a recursive backtracking algorithm to find the minimum number of transactions.  We will iterate
        through each person and try to settle their debt with every other person.  If the debt is settled, we will move
        on to the next person.  If the debt is not settled, we will try to settle the debt with another person.

        COMPLEXITY
        ----------

        Time complexity is O(2^n) because we are iterating through all possible transactions.

        Space complexity is O(n) because we are using recursion to solve the problem.
        """
        # Create a map of person -> amount.
        map: dict[int, int] = {}

        # First, figure out the balance that each person holds.
        for transaction in transactions:
            [from_person, to_person, amount] = transaction
            map[from_person] = map.get(from_person, 0) - amount
            map[to_person] = map.get(to_person, 0) + amount

        # Second, figure out all non-zero balances.  If we needed to know who is paying whom to achieve the minimum
        # number of transactions, we will need to have an array indexed by person.  However, we don't need to know that.
        #
        # We just need to know the number of transactions, so we can just get a list of non-zero balances and try to
        # zero them out.
        balances: list[int] = []
        for value in map.values():
            if value != 0:
                balances.append(value)

        # To settle debts efficiently, we should should match opposite balances together, and have them settle with each
        # other.  For example, if somebody has +5 balance, we should match then up with a -5 balance.  Matching two
        # balances with the same sign together will just result in extra transactions.
        #
        # Therefore, we should examine sequences of pairs that work towards an all zero balance sheet.  However,
        # different sequences of pairs could result in different numbers of transactions.  So we will need to examine
        # *all* possible pairs and keep track of the minimum tx's across all sequences.
        #
        # To do this effectively we need to use a recursive backtracking technique by trying to settle the ith balance
        # with i+1, i+2, i+3, etc and seeing how many tx's we incur at each step.
        def settle(left: int):
            # Skip over the already settled balances; if we've skipped over all balances, the minimum number of tx's is
            # zero.
            while left < len(balances) and balances[left] == 0:
                left += 1

            if left == len(balances):
                return 0

            # Try to settle the ith balance with the first opposite signed balance.
            min_transactions = math.inf
            for right in range(left + 1, len(balances)):
                # These balances were either both positive, or both negative, so do not bother executing a transaction
                # with them.
                if balances[left] * balances[right] > 0:
                    continue

                # Attempt to settle (i, j) by adding the ith balance to the jth balance.  Technically we don't have to
                # decrement balance[i] at all, since the next call to settle will start at j (and i will be skipped).
                # But it is included here for clarity.
                #
                # Note that we also do not have to ensure that we only add the minimum amount to balance.  For example,
                # if # balances[i] = 10 and balances[j] = -5, we don't need to only add 5 to balances[j].  As we
                # recursively move down the list, balances[j] will eventually zero out (and a more efficient path may be
                # discovered).
                #
                # Also, we are guaranteed that the balances *do* eventually zero out because there isn't any extra money
                # in the system.
                value = balances[left]
                balances[right] += value
                balances[left] = 0

                # The minimum number of transactions, if we've settled the ith balance, is 1 + the minimum after
                # settling the i+1th balance.
                #
                # As a further optimization we can use dynamic programming to avoid redundant calculations.  To do so
                # we'll need to use a bit mask to keep track of balance state.
                min_transactions = min(min_transactions, 1 + settle(left + 1))

                # Backtrack by undoing the transaction.
                balances[right] -= value
                balances[left] = value

            return min_transactions

        return int(settle(0))
