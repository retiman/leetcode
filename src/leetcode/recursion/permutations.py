# DIFFICULTY: MEDIUM
# ------------------
#
# Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.
#
# See https://leetcode.com/problems/permutations
class Solution:
    def permute(self, nums: list[int]) -> list[list[int]]:
        """
        SOLUTION
        --------

        To do this, recursively generate all permutations of elements without the head.  Then take all permutations
        created and add x to the front of the list.  This should result in n! number of arrays.

        The general strategy is to split the list into a head and a tail.  Generate all the permutations of the list
        without the head element, then put the head in front of each of the generated permutations.  Do this recursively
        for each element in the list will give all permutations.

        COMPLEXITY
        ----------

        Time complexity is O(n!) where n is the number of elements in the list.

        Space complexity is O(n) because we are storing the permutations in stack frames.
        """
        if not nums:
            return []

        if len(nums) == 1:
            return [nums]

        result: list[list[int]] = []
        for i, x in enumerate(nums):
            # Generate all permutations of the list without x.
            without = nums[:i] + nums[i + 1 :]
            ps = self.permute(without)

            # For each of the permutations, put the ith element in front of the permutations.
            for p in ps:
                result.append([x] + p)

        return result
