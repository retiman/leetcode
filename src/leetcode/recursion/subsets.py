# DIFFICULTY: MEDIUM
# ------------------
#
# Given an integer array nums of unique elements, return all possible subsets  (the power set).
#
# The solution set must not contain duplicate subsets. Return the solution in any order.
#
# See https://leetcode.com/problems/subsets
class Solution:
    def subsets(self, nums: list[int]) -> list[list[int]]:
        """
        SOLUTION
        --------
        To do this, recursively generate all combinations of elements of the set `xs - x`, then take all the subsets
        created and add x to them.  This should result in 2^n number of subsets.

        COMPLEXITY
        ----------
        Time complexity is O(2^n) where n is the number of elements in the list.

        Space complexity is O(n) because we are storing the subsets in stack frames.
        """

        def generate(xs: set[int]) -> set[set[int]]:
            if len(xs) == 0:
                return set(set())

            # Find the first element of xs and call it x; remove it from the set.
            x = xs.pop()

            # Make a copy of xs without x and generate all subsets of that set.
            rest = set(xs)
            excludeds = generate(rest)

            # Now generate all subsets with x by adding x to each excluded subset.
            includeds: set[set[int]] = set()
            for excluded in excludeds:
                included = set(excluded)
                included.add(x)
                includeds.add(included)

            # Combine the excluded and included subsets.
            return set.union(excludeds, includeds)

        result = generate(set(nums))
        return [list(subset) for subset in result]
