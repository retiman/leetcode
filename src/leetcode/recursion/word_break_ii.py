# DIFFICULTY: HARD
#
# Given a string s and a dictionary of strings wordDict, add spaces in s to construct a sentence where each word is a
# valid dictionary word. Return all such possible sentences in any order.
#
# Note that the same word in the dictionary may be reused multiple times in the segmentation.
#
# See https://leetcode.com/problems/word-break-ii
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> list[str]:
        """
        SOLUTION
        --------

        To do this, we can use a recursive backtracking algorithm.

        COMPLEXITY
        ----------

        Time complexity is O(2^n) where n is the number of words in the string.

        Space complexity is O(n) because we are storing the words in stack frames.
        """
        uniques = set(wordDict)
        memo: dict[int, list[str]] = {}

        # This generates all possible sentences using words that start at index i (if any exist) recursively.
        #
        # Since the only words we have available to us are the ones in the wordDict, we'll just have to loop through
        # from i to the end of the string to see if any of those substrings form a 'word'.
        def generate(i: int) -> list[str]:
            # Because we are slicing from [i, j), and slice is exclusive on the second index, we will be starting at
            # i + 1.  Therefore, do the bounds check for i == len(s).
            if i == len(s):
                return []

            if i in memo:
                return memo[i]

            sentences: list[str] = []
            # Attempt to generate sentences beginning with the 'word' from i to j.  Note that we want to slice from i to
            # j, so we actually want j to start at i + 1, and we also want to include j === s.length, because the slice
            # function is inclusive for the first index, and exclusive for the second index.
            for j in range(i + 1, len(s) + 1):
                word = s[i:j]
                if word not in uniques:
                    continue

                # Now generate all possible sentences using words that start at index j (if any exist), and put the word
                # from index i in front.
                rest = generate(j)

                # If the problem allowed us to generate sentences using a partial selection of characters (e.g. if we
                # couldn't make more words at this point), we could just add the current word to the result, even if
                # there were unused characters remaining.
                #
                # However, the problem does not allow us to do this, so we only add the current word if there are no
                # unused characters remaining.
                if len(rest) == 0 and j == len(s):
                    sentences.append(word)
                    continue

                # If we could generate words starting at j, prepend the word starting at index i.
                for sentence in rest:
                    sentences.append(word + " " + sentence)

            memo[i] = sentences
            return sentences

        return generate(0)
