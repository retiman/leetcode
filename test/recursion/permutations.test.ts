import { permute } from '../../src/recursion/permutations';

describe('permute', () => {
  test('permute - test case 1', async () => {
    expect(permute([])).toStrictEqual([[]]);
  });

  test('permute - test case 2', async () => {
    expect(permute([1, 2, 3])).toMatchSnapshot();
  });

  test('permute - test case 3', async () => {
    expect(permute([0, 1])).toMatchSnapshot();
  });

  test('permute - test case 4', async () => {
    expect(permute([1])).toStrictEqual([[1]]);
  });
});
