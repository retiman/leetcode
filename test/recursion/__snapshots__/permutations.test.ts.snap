// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permute permute - test case 2 1`] = `
[
  [
    1,
    2,
    3,
  ],
  [
    1,
    3,
    2,
  ],
  [
    2,
    1,
    3,
  ],
  [
    2,
    3,
    1,
  ],
  [
    3,
    1,
    2,
  ],
  [
    3,
    2,
    1,
  ],
]
`;

exports[`permute permute - test case 3 1`] = `
[
  [
    0,
    1,
  ],
  [
    1,
    0,
  ],
]
`;
